src/nvim/ex_cmds.c

```diff
@@ -5864,3 +5928,200 @@ void set_context_in_sign_cmd(expand_T *xp, char_u *arg)
+    long pat_size = STRLEN(pat);
+
+    // Get the width of the column which display the number of the line
+    linenr_T highest_num_line = 0;
+    kl_iter(MatchedLine, lmatch, current) {
+      highest_num_line = (*current)->data.lnum;
+    }
+
+    // computing the length of the column that will display line number
+    int col_width = log10(highest_num_line) + 1 + 3;
+
+    // will be allocated in the loop
+    char *str = NULL;
+
+    size_t old_line_size = 0;
+    size_t line_size;
+
+    // Append the lines to our buffer
+    kl_iter(MatchedLine, lmatch, current) {
+      MatchedLine mat = (*current)->data;
+      line_size = STRLEN(mat.line) + col_width + 1;
+
+      // Reallocation if str not long enough
+      if (line_size > old_line_size) {
+        str = xrealloc(str, line_size * sizeof(char));
+        old_line_size = line_size;
+      }
+
+      // put ' [ lnum]line' into str and append it to the incsubstitute buffer
+      snprintf(str, line_size, " [%*ld]%s", col_width - 3, mat.lnum, mat.line);
+      ml_append(line++, (char_u *)str, (colnr_T)line_size, false);
+
+      // highlight the replaced part
+      if (sub_size > 0) {
+        int i = 0;
+        int hlgroup_ls = syn_check_group((char_u *)"IncSubstitute", 13);
+
+        kl_iter(colnr_T, mat.start_col, col) {
+          src_id_highlight =
+          bufhl_add_hl(curbuf,
+                       src_id_highlight,
+                       hlgroup_ls,  // id of our highlight
+                       line,
+                       (*col)->data + col_width + i * (sub_size - pat_size) + 1,
+                       (*col)->data + col_width + i * (sub_size - pat_size) +
+                       sub_size);
+        }
+        i++;
+      }
+    }
+    xfree(str);
+  }
+
+  redraw_later(SOME_VALID);
+
+  // Restore the old window
+  win_enter(oldwin, false);
+  win_size_restore(&winsizes);
+  ga_clear(&winsizes);
+  exmode_active = save_exmode;
+  restart_edit = save_restart_edit;
+  cmdmsg_rl = save_cmdmsg_rl;
+  State = save_State;
+
+  char_u typestr[2];
+  typestr[0] = cmdwin_type;
+  typestr[1] = NUL;
+  apply_autocmds(EVENT_CMDWINLEAVE, typestr, typestr, false, curbuf);
+
+  cmdwin_type = 0;
+  int save_rd = RedrawingDisabled;
+  RedrawingDisabled = 0;
+  update_screen(0);
+  RedrawingDisabled = save_rd;
+
+  setmouse();
+}
+
+
+/// :substitute command implementation
+///
+/// Uses do_sub() to do the actual substitution. Calls :u after each typed
+/// character to undo the substitution unless finishing the command. If
+/// ics is set to 0, it just calls do_sub().
+void do_inc_sub(exarg_T *eap)
+{
+  // if incsubstitute disabled, do it the classical way
+  if (*p_ics == NUL || !eap->is_live) {
+    do_sub(eap);
+    return;
+  }
+
+  // Save the state of eap
+  char_u *tmp = eap->arg;
+
+  // Highlight the word and open the split
+  save_search_patterns();
+
+  // save the value of undolevels, we need to set it to maximum
+  // during incsub b/c we undo and redo a lot
```

Comment:
"a lot" is a bit misleading, it only uses one undolevel at a time. Maybe "set it to maximum to not inadvertently remove history in case undolevels had a low value".